pkg/assembler/graphql/resolvers/certifyBad.resolvers_test.go

```diff
@@ -0,0 +1,313 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolvers_test
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/guacsec/guac/internal/testing/ptrfrom"
+	"github.com/guacsec/guac/pkg/assembler/graphql/resolvers"
+	"github.com/guacsec/guac/pkg/assembler/helpers/expected"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/guacsec/guac/pkg/assembler/backends/inmem"
+	"github.com/guacsec/guac/pkg/assembler/graphql/model"
+)
+
+func TestCertifyBad(t *testing.T) {
+	type call struct {
+		Sub   model.PackageSourceOrArtifactInput
+		Match model.MatchFlags
+		CB    *model.CertifyBadInputSpec
+	}
+	tests := []struct {
+		Name         string
+		InPkg        []*model.PkgInputSpec
+		InSrc        []*model.SourceInputSpec
+		InArt        []*model.ArtifactInputSpec
+		Calls        []call
+		Query        *model.CertifyBadSpec
+		ExpCB        []*model.CertifyBad
+		ExpIngestErr bool
+		ExpQueryErr  bool
+	}{
+		{
+			Name:  "Ingest with two subjects",
+			InSrc: []*model.SourceInputSpec{expected.S1},
+			InArt: []*model.ArtifactInputSpec{expected.A1},
+			Calls: []call{
+				{
+					Sub: model.PackageSourceOrArtifactInput{
+						Source:   expected.S1,
+						Artifact: expected.A1,
+					},
+					CB: &model.CertifyBadInputSpec{
+						Justification: "test justification",
+					},
+				},
+			},
+			ExpIngestErr: true,
+		},
+		{
+			Name:  "Query with two subjects",
+			InSrc: []*model.SourceInputSpec{expected.S1},
+			Calls: []call{
+				{
+					Sub: model.PackageSourceOrArtifactInput{
+						Source: expected.S1,
+					},
+					CB: &model.CertifyBadInputSpec{
+						Justification: "test justification",
+					},
+				},
+			},
+			Query: &model.CertifyBadSpec{
+				Subject: &model.PackageSourceOrArtifactSpec{
+					Package: &model.PkgSpec{
+						Version: ptrfrom.String("2.11.1"),
+					},
+					Artifact: &model.ArtifactSpec{
+						Algorithm: ptrfrom.String("asdf"),
+					},
+				},
+			},
+			ExpQueryErr: true,
+		},
+	}
+	ignoreID := cmp.FilterPath(func(p cmp.Path) bool {
+		return strings.Compare(".ID", p[len(p)-1].String()) == 0
+	}, cmp.Ignore())
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.Name, func(t *testing.T) {
+			b, err := inmem.GetBackend(nil)
```

Comment:
Shouldn't be using the inmem backend here. Unit test should only test the package under test. Create a mock backend for the resolver to call.

Reply:
Awesome, thanks for the heads up about the mock usage.